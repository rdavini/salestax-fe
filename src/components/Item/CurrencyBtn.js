import React, { useEffect, useState } from 'react'
import { Dropdown, DropdownButton } from 'react-bootstrap';
import api from '../../services/api'

const CurrencyBtn = ({items, setItems, currency, setCurrency}) => {
    const [eurToInrRate, setEurToInrRate] = useState(null)
    const [eurToUsdRate, setEurToUsdRate] = useState(null)
    const [usdToEurRate, setUsdToEurRate] = useState(null)
    const [usdToInrRate, setUsdToInrRate] = useState(null)
    const [inrToEurRate, setInrToEurRate] = useState(null)
    const [inrToUsdRate, setInrToUsdRate] = useState(null)

    useEffect(()=>{
        // EUR -> INR, USD
        api.api_fixer_latest_rate('EUR', ['INR', 'USD']).then((res)=>{
            setEurToInrRate( res.data.rates.INR )
            setEurToUsdRate( res.data.rates.USD )
        }).catch((err)=> console.log(err))

        //USD -> EUR, INR
        api.api_fixer_latest_rate('USD', ['EUR', 'INR']).then((res)=>{
            setUsdToEurRate( res.data.rates.EUR )
            setUsdToInrRate( res.data.rates.INR )
        }).catch((err)=> console.log(err))

        //INR -> EUR, USD
        api.api_fixer_latest_rate('INR', ['EUR', 'USD']).then((res)=>{
            setInrToEurRate( res.data.rates.EUR )
            setInrToUsdRate( res.data.rates.USD )
        }).catch((err)=> console.log(err))
    }, [])

    const handleSubmit = (new_currency) =>{
        let temp_items = []
        let rate = 1;

        switch(new_currency){
            case 'EUR':
                rate = currency == 'INR' ? inrToEurRate : usdToEurRate
                break
            case 'INR':
                rate = currency == 'EUR' ? eurToInrRate : usdToInrRate
                break
            case 'USD':
                rate = currency == 'INR' ? inrToUsdRate : eurToUsdRate
                break
        }
        items.forEach(item =>{
            temp_items.push({
                desc: item.desc,
                price: item.price*rate,
                qty: item.qty
            })
        })

        setItems(temp_items)
        setCurrency(new_currency)
    }

    return (
      <DropdownButton id="dropdown-basic-button" variant="light" title={currency} style={{marginTop: '1.5rem'}}>
        {currency == 'EUR' && 
        <div>
            <Dropdown.Item href="#/action-2" onClick={()=>handleSubmit('INR')}>INR</Dropdown.Item>
            <Dropdown.Item href="#/action-3" onClick={()=>handleSubmit('USD')}>USD</Dropdown.Item>
        </div>
        }
        {currency == 'USD' && 
        <div>
            <Dropdown.Item href="#/action-2" onClick={()=>handleSubmit('INR')}>INR</Dropdown.Item>
            <Dropdown.Item href="#/action-3" onClick={()=>handleSubmit('EUR')}>EUR</Dropdown.Item>
        </div>
        }
        {currency == 'INR' && 
        <div>
            <Dropdown.Item href="#/action-2" onClick={()=>handleSubmit('EUR')}>EUR</Dropdown.Item>
            <Dropdown.Item href="#/action-3" onClick={()=>handleSubmit('USD')}>USD</Dropdown.Item>
        </div>
        }
      </DropdownButton>
    );
}

export default CurrencyBtn