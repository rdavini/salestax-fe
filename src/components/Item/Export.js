import React from 'react'
import PropTypes from 'prop-types';

const Export = ({header, items, totalCost, totalTax, currency}) => {
    const downloadTxtFile = () => {
        let exp_data = []
        header.forEach((h)=> exp_data.push((h.title+" ")))
        exp_data.push("\n")
        items.forEach((it) => {
            header.forEach((index) =>{
                exp_data.push(it[index.prop]+ " ")
            })
            exp_data.push(" ("+currency+")")
            exp_data.push("\n")
        })

        exp_data.push("Total Tax ")
        exp_data.push(totalTax)
        exp_data.push(" ("+currency+")")
        exp_data.push("\n")
        exp_data.push("Total ")
        exp_data.push(totalCost)
        exp_data.push(" ("+currency+")")


        const element = document.createElement("a");
        const file = new Blob(exp_data, {type: 'text/plain'});
        element.href = URL.createObjectURL(file);
        element.download = "salestax.txt";
        document.body.appendChild(element); // Required for this to work in FireFox
        element.click();
    }
    
    return <button onClick={downloadTxtFile} className='btn btn-light' style={{display: 'block'}}>Download</button>
}

Export.propTypes = {
    header: PropTypes.array.isRequired,
    items: PropTypes.array.isRequired,
    totalCost: PropTypes.number.isRequired,
    totalTax: PropTypes.number.isRequired,
    currency: PropTypes.string.isRequired,
};

export default Export