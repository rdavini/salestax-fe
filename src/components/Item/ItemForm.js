import React, { useState, useEffect } from 'react'
import { Col, Row, Form} from 'react-bootstrap';
import PropTypes from 'prop-types';

const ItemForm = ({itemsPayload, setItemsPayload, index}) => {  
    const [qty, setQty] = useState('')
    const [desc, setDesc] = useState('')
    const [price, setPrice] = useState('')

    const styles={
      floatL: {
        float: 'left'
      }
    }

    useEffect(()=>{
        setItemsPayload(itemsPayload.concat({qty: '', desc: '', price: ''}))
    }, [])

    useEffect(()=>{
        let new_row = {qty: qty, desc: desc, price: price}
        
        let temp = itemsPayload;
        temp[index] = new_row
        setItemsPayload(temp)
    }, [qty, desc, price])

    return <Row>
    <Col>
    <Form.Group controlId="formBasicQuantity">
      <Form.Label style={styles.floatL}>Qty</Form.Label>
      <Form.Control type="number" min="1" step="1" className='form-control' value={qty} onChange= { e => setQty(e.target.value)} />
    </Form.Group>
    </Col>
    <Col>
      <Form.Group controlId="formBasicDescription">
        <Form.Label style={styles.floatL}>Item Description</Form.Label>
        <Form.Control type="text" className='form-control' placeholder="Description" value={desc} onChange= { e => setDesc(e.target.value)} />
      </Form.Group>
    </Col>
    <Col>
      <Form.Group controlId="formBasicPrice">
        <Form.Label style={styles.floatL}>Shelf Price</Form.Label>
        <Form.Control type="number" min="0.01" step="0.01" className='form-control' placeholder="R$" value={price} onChange= { e => setPrice(e.target.value)}/>
      </Form.Group>
    </Col>
  </Row>
}

ItemForm.propTypes = {
  itemsPayload: PropTypes.array.isRequired,
  setItemsPayload: PropTypes.func.isRequired,
  index: PropTypes.number.isRequired
};

export default ItemForm