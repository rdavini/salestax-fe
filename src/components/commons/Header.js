import React from 'react'

const Header = () => {
    const styles ={
        header: {
            borderBottom: 1,
            borderBottomStyle: 'solid',
            borderColor: '#CBD5DE',
            backgroundColor: '#FFFFFF'
        }
    }
    return <div style={styles.header}>
            <h2 textalign='center'>Sales Tax Calculator</h2>
        </div>
}

export default Header