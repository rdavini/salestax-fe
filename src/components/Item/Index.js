import React, {useEffect, useState} from 'react';
import api from '../../services/api';
import CurrencyBtn from './CurrencyBtn';
import Export from './Export';
import Table from './Table'
import useCurrency from '../../hooks/useCurrency'
import PropTypes from 'prop-types';

const styles = {
    back: {
        float: 'left',
        color: 'blue'
    },
    title: {
        textAlign: 'left',
        float: 'left',
        marginRight: 5
    }
}

const headers = [
    { title: 'Qty', prop: 'qty' },
    { title: 'Item Description', prop: 'desc' },
    { title: 'Price', prop: 'price' }
  ];

const Index = ({setUrl}) => {
    const [items, setItems] = useState([])
    const [totalCost, setTotalCost] = useState()
    const [totalTax, setTotalTax] = useState()
    const { currency, setCurrency, currencyIcon} = useCurrency()
    const [error, setError] = useState(null)
    const [rate, setRate] = useState(1)

    const getItems = () => {
        api.api_sales_tax_get_items().then(rs =>{
            setItems(rs.data)
        }).catch((err)=> {
            console.log(err)
            setError(err.response.message)
        })
    }

    const getTotalInfo = () => {
        api.api_sales_tax_total_info().then(rs => {
            setTotalCost( rs.data.total_cost )
            setTotalTax( rs.data.total_tax )
        }).catch((err)=>{
            console.log(err)
            setError(err.response.message)
        })
    }

    useEffect(()=>{
        if(rate != 1){
            setTotalCost((totalCost*rate).toFixed(2))
            setTotalTax((totalTax*rate).toFixed(2))
        }
    }, [rate])

    useEffect(()=>{
        getItems();
        getTotalInfo();
    }, [])

    return <div className='container col-md-8' style={{marginTop: 50}}>
            { error  &&
                <div className="alert alert-danger" role="alert">
                    { error }
                </div>
            }
            <br/>
            {
                items.length > 0 ?
                <div>
                    <h2 style={styles.title}>Receipt</h2>
                    <Export header={headers} items={items} totalCost={totalCost} totalTax={totalTax} currency={currency} />
                    <CurrencyBtn items={items} setItems={setItems} currency={currency} setCurrency={setCurrency} setError={setError} rate={rate} setRate={setRate}/>
                    <Table items={items} totalCost={totalCost} totalTax={totalTax} currencyIcon={currencyIcon} rate={rate}/>
                </div>
                :
                'Loading items ...'
            }
            <span style={styles.back} onClick={()=> setUrl('create')}>back</span>
        </div>
}

Index.propTypes = {
    setUrl: PropTypes.func.isRequired,
};

export default Index