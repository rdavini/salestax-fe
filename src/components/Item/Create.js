import React, { useEffect, useState } from 'react'
import { Button, Form} from 'react-bootstrap';
import api from '../../services/api'
import ItemForm from './ItemForm';
import PropTypes from 'prop-types';

const Create = ({setUrl}) => {
  const [itemsPayload, setItemsPayload] = useState([{qty: '', desc: '', price: ''}])
  const [itemsRows, setItemsRows] = useState([])
  const [index, setIndex] = useState(0)
  const [error, setError] = useState(null)

  const styles= {
    contentMarginTop: {
      marginTop: 50
    },
    floatL:{
      float: 'left'
    },
    alignL:{
      textAlign: 'left'
    },
    btnMarginTop:{
      marginTop: 30
    }
  }

  useEffect(()=>{
    setItemsRows([<ItemForm itemsPayload={itemsPayload} setItemsPayload={setItemsPayload} index={index} />])
  }, [])

  const handleSubmit = () => {
    api.api_sales_tax_create_items(itemsPayload).then( () => {
      setUrl('index')
    }).catch((err) => {
      console.log(err)
      setError(err.response.data)
    });
  }

  const addRow = () => {
    setItemsRows(itemsRows.concat(<ItemForm itemsPayload={itemsPayload} setItemsPayload={setItemsPayload} index={index+1}/>))
    setIndex(index+1)
  }

  return (
    <div className='container' style={styles.contentMarginTop}>
       { error  &&
        <div className="alert alert-danger" role="alert">
            { error }
        </div> 
      }

      <h2 style={styles.alignL}>Item Details</h2>
      <Form>
        <div className='col-md-10' style={styles.floatL}>
          {itemsRows}
        </div>

        <Button style={styles.btnMarginTop} onClick={addRow}>+</Button>
        <div style={{clear: 'both', paddingTop: 20}}>
          <Button  variant="primary" onClick={handleSubmit}>
            Generate Receipt
          </Button>
        </div>
      </Form>
    </div>
  );
}

Create.propTypes = {
  setUrl: PropTypes.func.isRequired,
};

export default Create;