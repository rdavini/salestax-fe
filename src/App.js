import logo from './logo.svg';
import './App.css';
import 'bootstrap/dist/css/bootstrap.min.css';
import Item from './components/Item/Item';
import Header from './components/commons/Header';

function App() {
  return (
    <div className="App">
      <Header />
      <Item />
    </div>
  );
}

export default App;
