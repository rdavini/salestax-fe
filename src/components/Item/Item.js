import React from 'react'
import Create from './Create'

const Item = () => {
    return <Create/>
}

export default Item