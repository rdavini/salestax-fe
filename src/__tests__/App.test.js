import React from 'react';
import App from '../App';
import { create } from 'react-test-renderer'
import Header from '../components/commons/Header'
import Item from '../components/Item/Item'
import { render, screen } from '@testing-library/react'

describe('Elements present test',()=>{
    test('testing snapshot', () => {
        let tree = create(<App />)
        expect(tree.toJSON()).toMatchSnapshot();
    })
    
    test('renders header', () => {
        render(<Header />);
        const linkElement = screen.getByText(/Sales Tax Calculator/i);
        expect(linkElement).toBeInTheDocument();
    });

    test('renders Item', () => {
        render(<Item />);
        
        // title
        const linkElement = screen.getByText(/Item Details/i);
        expect(linkElement).toBeInTheDocument();

        // qty label
        const qtyLabel = screen.getByText(/Qty/i)
        expect(qtyLabel).toBeInTheDocument()
        
        
        // itemDesc label
        const itemDescLabel = screen.getByText(/Item Description/i)
        expect(itemDescLabel).toBeInTheDocument()
        
        //itemDesc input text
        const itemDescInput = screen.getByPlaceholderText('Description');
        expect(itemDescInput).toBeInTheDocument()
        
        // price label
        const priceLabel = screen.getByText(/Shelf Price/i)
        expect(priceLabel).toBeInTheDocument()

        //price input text
        const priceInput = screen.getByPlaceholderText('R$');
        expect(priceInput).toBeInTheDocument()

        //add btn
        const btnPlus = screen.getByRole('button', { name: '+' });
        expect(btnPlus).toBeInTheDocument()
        
        //genReceipt btn
        const btnGenReceipt = screen.getByRole('button', { name: 'Generate Receipt' });
        expect(btnGenReceipt).toBeInTheDocument()

    });
      
})