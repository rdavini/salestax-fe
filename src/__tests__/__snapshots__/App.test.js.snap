// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Elements present test testing snapshot 1`] = `
<div
  className="App"
>
  <div
    style={
      Object {
        "backgroundColor": "#FFFFFF",
        "borderBottom": 1,
        "borderBottomStyle": "solid",
        "borderColor": "#CBD5DE",
      }
    }
  >
    <h2
      textalign="center"
    >
      Sales Tax Calculator
    </h2>
  </div>
  <div
    className="container"
    style={
      Object {
        "marginTop": 50,
      }
    }
  >
    <h2
      style={
        Object {
          "textAlign": "left",
        }
      }
    >
      Item Details
    </h2>
    <form
      className=""
    >
      <div
        className="col-md-10"
        style={
          Object {
            "float": "left",
          }
        }
      />
      <button
        className="btn btn-primary"
        disabled={false}
        onClick={[Function]}
        style={
          Object {
            "marginTop": 30,
          }
        }
        type="button"
      >
        +
      </button>
      <div
        style={
          Object {
            "clear": "both",
            "paddingTop": 20,
          }
        }
      >
        <button
          className="btn btn-primary"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Generate Receipt
        </button>
      </div>
    </form>
  </div>
</div>
`;
